File: lib/haveapi/client/authentication/token.rb
module HaveAPI::Client::Authentication
  class Token < Base
    register :token
    attr_reader :token, :valid_to

    def setup
      @via = @opts[:via] || :header
      @token = @opts[:token]
      @valid_to = @opts[:valid_to]

      request_token unless @token

      @configured = true
    end

    def request_url_params
      return {} unless @configured
      check_validity
      @via == :query_param ? {@desc[:query_parameter] => @token} : {}
    end

    def request_headers
      return {} unless @configured
      check_validity
      @via == :header ? {@desc[:http_header] => @token} : {}
    end

    def save
      {token: @token, valid_to: @valid_to}
    end

    def load(hash)
      @token = hash[:token]
      @valid_to = hash[:valid_to]
    end

    protected
    def request_token
      a = HaveAPI::Client::Action.new(@communicator, :request, @desc[:resources][:token][:actions][:request], [])
      ret = a.execute({
                          login: @opts[:user],
                          password: @opts[:password],
                          lifetime: @opts[:lifetime],
                          interval: @opts[:interval] || 300})

      raise AuthenticationFailed.new('bad username or password') unless ret[:status]

      @token = ret[:response][:token][:token]

      @valid_to = ret[:response][:token][:valid_to]
      @valid_to = @valid_to && DateTime.iso8601(@valid_to).to_time
    end

    def check_validity
      if @valid_to && @valid_to < Time.now
        if @opts[:user] && @opts[:password]
          request_token
        else
          raise AuthenticationFailed.new('token expired')
        end
      end
    end
  end
end

File: lib/haveapi/client/authentication/token.ex
defmodule HaveAPI.Client.Authentication.Token do
  alias HaveAPI.Client

  @behaviour Client.Authentication

  def setup(conn, opts) do
    if opts[:token] do
      {:ok, %{token: opts[:token], via: opts[:via] || :header}}

    else
      case request(token_conn(conn), opts) do
        {:ok, token, valid_to} ->
          {:ok, %{token: token, valid_to: valid_to, via: opts[:via] || :header}}

        {:error, msg} ->
          {:error, msg}
      end
    end
  end

  def authenticate(req, opts) do
    do_authenticate(req, opts[:via], opts[:token])
  end

  def request(conn, opts) do
    res = Client.call(conn, "request", input: %{
      login: opts[:username],
      password: opts[:password],
      lifetime: opts[:lifetime] || "renewable_auto",
      interval: opts[:interval] || 300
    })

    if Client.Response.ok?(res) do
      params = Client.Response.params(res)

      {:ok, params["token"], params["valid_to"]}

    else
      {:error, res.message}
    end
  end

  defp do_authenticate(req, :header, token) do
    Client.Request.add_header(req, desc(req.conn)["http_header"], token)
  end

  defp do_authenticate(req, :query_parameter, token) do
    Client.Request.add_query_param(req, desc(req.conn)["query_parameter"], token)
  end

  defp token_conn(conn) do
    %{conn |
      resource_path: ["token"],
      resource_desc: desc(conn)["resources"]["token"],
      path_params: []
    }
  end

  defp desc(conn) do
    conn.description["authentication"]["token"]
  end
end
